### PR TITLE
Split the `When_distributing_event` test

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@andreasohlund @yvesgoeleven please take a look.

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        public async Task Each_event_should_be_delivered_to_single_instance_of_each_subscriber()
+        {
+            await Scenario.Define<Context>()
+                .WithEndpoint<Publisher>(b => b.When(c => c.EndpointsStarted, async (session, c) => { await session.Publish(new MyEvent()); }))
+                .WithEndpoint<SubscriberA_1>()
+                .WithEndpoint<SubscriberA_2>()
+                .WithEndpoint<SubscriberB_1>()
+                .WithEndpoint<SubscriberB_2>()
+                .Done(c => c.SubscriberACounter > 0 && c.SubscriberBCounter > 0)
+                .Repeat(r => r.For<AllTransportsWithCentralizedPubSubSupport>())
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs]: Sidenote: Should we rename this to `AllTransportsWithMulticastSupport` ?

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
+﻿namespace NServiceBus.AcceptanceTests.Routing
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus.AcceptanceTesting;
+    using NServiceBus.AcceptanceTests.EndpointTemplates;
+    using NServiceBus.AcceptanceTests.ScenarioDescriptors;
+    using NUnit.Framework;
+
+    public class When_publishing_to_scaled_out_subscribers_with_native_pubsub : NServiceBusAcceptanceTest
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs]: how about: `When_publishing_to_scaled_out_subscribers_on_multicast_transports`?


Or should we have a `/Rounting/Multicast` folder instead?

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_storage_driven_pubsub.cs [hunk 3]
```diff
@@ -4,56 +4,32 @@
     using System.Threading.Tasks;
     using NServiceBus.AcceptanceTesting;
     using NServiceBus.AcceptanceTests.EndpointTemplates;
+    using NServiceBus.AcceptanceTests.ScenarioDescriptors;
     using NServiceBus.Routing;
     using NUnit.Framework;
 
-    public class When_distributing_an_event : NServiceBusAcceptanceTest
+    public class When_publishing_to_scaled_out_subscribers_with_storage_driven_pubsub : NServiceBusAcceptanceTest
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_storage_driven_pubsub.cs]: Rename to "unicast"? (see above)

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        [Test]
+        public async Task Each_event_should_be_delivered_to_single_instance_of_each_subscriber()
+        {
+            await Scenario.Define<Context>()
+                .WithEndpoint<Publisher>(b => b.When(c => c.EndpointsStarted, async (session, c) => { await session.Publish(new MyEvent()); }))
+                .WithEndpoint<SubscriberA_1>()
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs]: would it work to use `.WithEndpoint<SubscriberA>()` twice? the WithEndpoint has an overload which would allow you to configure the instance discriminator.

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_storage_driven_pubsub.cs (hunk 3 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_storage_driven_pubsub.cs]: :+1:

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs (hunk 2 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs]: :+1:

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs (hunk 4 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs]: :+1: . Thanks for the hint. It simplified *a lot*.

Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_with_native_pubsub.cs]: glad it works :) looks really better :+1:

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_multicast_transports.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    using NServiceBus.AcceptanceTests.ScenarioDescriptors;
+    using NUnit.Framework;
+
+    public class When_publishing_to_scaled_out_subscribers_on_multicast_transports : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Each_event_should_be_delivered_to_single_instance_of_each_subscriber()
+        {
+            await Scenario.Define<Context>()
+                .WithEndpoint<Publisher>(b => b.When(c => c.EndpointsStarted, async (session, c) => { await session.Publish(new MyEvent()); }))
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_multicast_transports.cs]: is there no risk of a race condition with subscribers?

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                }))
+                .WithEndpoint<SubscriberA>(b => b.CustomConfig(c => c.ScaleOut().InstanceDiscriminator("1")))
+                .WithEndpoint<SubscriberA>(b => b.CustomConfig(c => c.ScaleOut().InstanceDiscriminator("2")))
+                .WithEndpoint<SubscriberB>(b => b.CustomConfig(c => c.ScaleOut().InstanceDiscriminator("1")))
+                .WithEndpoint<SubscriberB>(b => b.CustomConfig(c => c.ScaleOut().InstanceDiscriminator("2")))
+                .Done(c => c.ProcessedByA > 0 && c.ProcessedByB > 0)
+                .Repeat(r => r.For<AllTransportsWithMessageDrivenPubSub>())
+                .Should(c =>
+                {
+                    Assert.IsTrue(c.ProcessedByA == 1);
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Assert.AreEqual?

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    using NServiceBus.Routing;
+    using NUnit.Framework;
+
+    public class When_publishing_to_scaled_out_subscribers_on_unicast_transports : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Each_event_should_be_delivered_to_single_instance_of_each_subscriber()
+        {
+            await Scenario.Define<Context>()
+                .WithEndpoint<Publisher>(b => b.When(c => c.SubscribersCounter == 4, async (session, c) =>
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: In https://github.com/Particular/NServiceBus/pull/3459/files#diff-a81add4da1810ee4694342b94ff80cb1R15 you use `EndpointsStarted`. Should we align?

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_multicast_transports.cs (hunk 5 above)
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_multicast_transports.cs]: @timbussmann The ATT framework sets the `EndpointsStarted` flag after all endpoints are successfully started

Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_multicast_transports.cs]: ah, EndpointS, got it :+1:

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs (hunk 7 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Not sure if that will be enough to ensure they are subscribed. There might be a race condition.

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs (hunk 6 above)
Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: :+1:

**Comment by SzymonPobiega** [comment]
Fixed `Assert`s, rebased and squashed.

File: src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs (hunk 7 above)
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Which would be a bug in the ATT framework, right? And also mean that the other test would not work

Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Why? There is nothing in `EndpointsStarted` that promises that all subscription messages are processed and I don't see a way to ensure that in a generic way.

Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: >There is nothing in EndpointsStarted that promises that all subscription messages are processed and I don't see a way to ensure that in a generic way.

Agreed, and it shouldn't be either IMO - https://github.com/Particular/NServiceBus/issues/3439#issuecomment-181842709

Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: Let me rephrase. Why are you then using once the subscriber count approach and then the `EndpointStarted` approach https://github.com/Particular/NServiceBus/pull/3459#discussion-diff-52742348R15?

Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: On multicast transports subscribe happens in the FST and is synchronous so in this case I can be sure that when all endpoints are started, subscriptions are there.

Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: >On multicast transports subscribe happens in the FST and is synchronous

Can we really assume this? 

eg. exchange.bind in rabbit can be async if you use the `no-wait` flag?

http://www.rabbitmq.com/amqp-0-9-1-reference.html#exchange.bind.no-wait

Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: @andreasohlund I have a follow up discussion I want to have with you.
Elevator pitch: I think we should get rid of message driven pubsub entirely

Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: You know I like to kill things

Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: :fire: :fire: :fire: :fire: :fire: :fire:

Comment (SzymonPobiega) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: @andreasohlund that's the best thing we can do here. We have no insight into the fact if the routing topology has been updated.

Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/When_publishing_to_scaled_out_subscribers_on_unicast_transports.cs]: >that's the best thing we can do here. We have no insight into the fact if the routing topology has been updated.

Fair enough, was just pointing out that we can't tell the users that they are guaranteed to be subscribed. If they want to be 100% sure they need to script it ahead of time: https://github.com/Particular/NServiceBus/issues/3439#issuecomment-181842709

**Comment by danielmarbach** [comment]
Pulled in.
